Quickstart update for transforms (#4897)
PennyLaneAI/pennylane: doc/introduction/transforming_circuits.rst

.. role:: html(raw)
   :format: html

.. _intro_ref_transform_circuits:

Transforming circuits

> =====================

Reviewer: Yeah good question :thinking: One argument against merging this into "compiling circuits" could be that the result feels a bit long.

Could this content just be at the bottom of [`qml.transforms`](https://xanaduai-pennylane--4897.com.readthedocs.build/en/4897/code/qml_transforms.html)?

> =====================

Reviewer 2: Oh yeah, that's a great idea!

> =====================

Author: I agree with you I will move it to the transforms page, thanks!

A quantum transform is a crucial concept in PennyLane, it is represented by a function that takes a circuit as input and
yields one or more transformed circuits along with a post-processing function. The post-processing function is applied
to the results obtained after executing the transformed circuit(s). This becomes particularly valuable when a transform
generates multiple circuits, requiring a method to aggregate the results (e.g. parameter shift or Hamiltonian expansion).

> In PennyLane, these requirements are translated as follows:
> 
> * A transform accepts a :class:`~.QuantumTape` as its primary input and
>   returns a sequence of :class:`~.QuantumTape` and an associated processing function.
> 
> To streamline the creation of transforms and ensure their versatility across various circuit abstractions in PennyLane
> (:class:`~.QuantumTape`, :class:`~.QNode` and quantum functions), we have created a simple
> decorator :func:`pennylane.transform` that can be applied on your quantum transform respecting the above contract.
> 
> Creating your own transform
> ---------------------------

Reviewer 2: ```suggestion
Creating your own transform
---------------------------

To streamline the creation of transforms and ensure their versatility across
various circuit abstractions in PennyLane, the
:func:`pennylane.transform` is available.

This decorator registers transforms that accept a :class:`~.QuantumTape`
as its primary input, and returns a sequence of :class:`~.QuantumTape`
and an associated processing function.
```

To illustrate the process of creating a quantum transform, let's consider a straightforward example. Suppose we want
a transform that removes all :class:`~.RX` operations from a given circuit. In this case, we merely need to filter the
original :class:`~.QuantumTape` and return a new one without the filtered operations. As we don't require a specific processing
function in this scenario, we include a null function that simply returns the result.

.. code-block:: python

    from typing import Sequence, Callable
    from pennylane.tape import QuantumTape

    def remove_rx(tape: QuantumTape) -> (Sequence[QuantumTape], Callable):

        operations = filter(lambda op: op.name != "RX", tape.operations)
        new_tape = type(tape)(operations, tape.measurements, shots=tape.shots)

        def null_postprocessing(results):

>             return results[0]

Reviewer 2: Why do you have to index? For me, this isn't a null function, I would expect `lambda *args: args` to be a null function 🤔

>         return [new_tape], null_postprocessing

Reviewer 2: A feature request: I would love if I could just do `return [new_tape], None` and have it automatically use a null processing function.

I would also love if I could do:

```python
return new_tape, None
or?
return new_tape
```

for a single circuit->circuit transform.

Author: I think that would be a nice improvement for users indeed. Could we add that to the plan for Q1 @trbromley ?

        def null_postprocessing(results):
            return qml.sum(results)

        return [tape, shifted_tape], null_postprocessing

Composability of transforms
---------------------------

Transforms are inherently composable on :class:`~.QNode`, meaning that transforms with compatible post-processing
functions can be successively applied to QNodes. For example, this allows for the application of multiple compilation
passes on a QNode to maximize gate reduction before execution.

.. code-block:: python

        dev = qml.device("default.qubit", wires=1)
        @qml.merge_rotations
        @qml.cancel_inverses
        @qml.qnode(device=dev):
        def circuit(x, y):
            qml.Hadamard(wires=0)
            qml.Hadamard(wires=0)
            qml.RX(x, wires=0)
            qml.RY(y, wires=0)
            qml.RZ(y, wires=0)
            qml.RY(x, wires=0)
            return qml.expval(qml.PauliZ(wires=0))

In this example, inverses are canceled, leading to the removal of two Hadamard gates. Subsequently, rotations are
merged into a single :class:`qml.Rot` gate. Consequently, two transforms are successfully applied to the circuit.

> Relevant links
> --------------

Reviewer 2: ```suggestion
Additional information
----------------------
```
